tests/apps/demultiplex/test_sample_models.py

```diff
@@ -0,0 +1,398 @@
+from unittest.mock import Mock
+
+import pytest
+
+from cg.apps.demultiplex.sample_sheet.index import (
+    INDEX_ONE_PAD_SEQUENCE,
+    INDEX_TWO_PAD_SEQUENCE,
+    LONG_INDEX_CYCLE_NR,
+    SHORT_SAMPLE_INDEX_LENGTH,
+    get_reverse_complement_dna_seq,
+)
+from cg.apps.demultiplex.sample_sheet.sample_models import (
+    FlowCellSampleBcl2Fastq,
+    FlowCellSampleBCLConvert,
+)
+from cg.constants.demultiplexing import IndexOverrideCycles, IndexSettings
+from cg.constants.symbols import DASH, EMPTY_STRING
+from cg.models.demultiplex.run_parameters import RunParameters
+
+
+@pytest.mark.parametrize(
+    "lims_sample",
+    [
+        {"index": "GTCTACAC-GCCAAGGT", "sample_id": "ACC123"},
+        {"lane": 1, "sample_id": "ACC123"},
+        {"lane": 1, "index": "GTCTACAC-GCCAAGGT"},
+    ],
+    ids=["no sample id", "no index", "no lane"],
+)
+def test_validate_inputs_bcl_convert_sample_missing_attribute(lims_sample: dict):
+    """Test that validating a BCLConvert sample with a missing attribute fails."""
+    # GIVEN a raw LIMS sample without a mandatory attribute
+
+    # WHEN validating the sample
+    with pytest.raises(AttributeError) as exc_info:
+        FlowCellSampleBCLConvert.validate_inputs(lims_sample=lims_sample)
+
+    # THEN a pydantic validation error is raised
+    assert "validate_inputs" in str(exc_info.value)
+
+
+@pytest.mark.parametrize(
+    "lims_index, expected_index_1, expected_index2",
+    [
+        ("CCATTCGANNNNNNNNN-GTTGTCCG", "CCATTCGA", "GTTGTCCG"),
+        ("CGATAGCAGG-AATGCTACGA", "CGATAGCAGG", "AATGCTACGA"),
+        ("ACAGCAAG", "ACAGCAAG", ""),
+        ("ACAGCAAG-", "ACAGCAAG", ""),
+        ("ATAGCAGG-AATGCTAC\n", "ATAGCAGG", "AATGCTAC"),
+    ],
+    ids=[
+        "17-nt index1",
+        "equal-length indexes",
+        "single index",
+        "single index with dash",
+        "index with newline",
+    ],
+)
+def test_separate_indexes_dual_run(lims_index: str, expected_index_1: str, expected_index2: str):
+    """Test that parsing different kinds of dual-run raw indexes as index and index2 works."""
+    # GIVEN a sample sheet with a single sample on one lane
+    sample = FlowCellSampleBCLConvert(lane=1, index=lims_index, sample_id="ACC123")
+
+    # WHEN separating the index
+    sample.separate_indexes(is_run_single_index=False)
+
+    # THEN the index should be separated
+    assert sample.index == expected_index_1
+    assert sample.index2 == expected_index2
+
+
+def test_separate_indexes_single_run(bcl_convert_flow_cell_sample: FlowCellSampleBCLConvert):
+    """Test index2 is ignored when parsing a double index in a single index run."""
+    # GIVEN a sample with a double index
+
+    # WHEN separating the index
+    bcl_convert_flow_cell_sample.separate_indexes(is_run_single_index=True)
+
+    # THEN the index should be separated
+    assert bcl_convert_flow_cell_sample.index == "GTCTACAC"
```

Comment:
Could we make this string into a fixture and use that fixture in the instantiation of bcl_convert_flow_cell_sample?

Reply:
Agreeing with Isak, and name it something descriptive

Reply:
fixed